diff --git a/rootdir/init.rc b/rootdir/init.rc
index d8e960a..20320cd 100644
--- a/rootdir/init.rc
+++ b/rootdir/init.rc
@@ -217,6 +217,7 @@ on post-fs-data
 
     # create basic filesystem structure
     mkdir /data/misc 01771 system misc
+    mkdir /data/misc/bluetoothd 0770 bluetooth bluetooth
     mkdir /data/misc/adb 02750 system shell
     mkdir /data/misc/audit 02750 audit system
     mkdir /data/misc/bluedroid 0770 bluetooth net_bt_stack
@@ -495,12 +496,12 @@ service netd /system/bin/netd
 service debuggerd /system/bin/debuggerd
     class main
 
-service ril-daemon /system/bin/rild
+service ril-daemon /system/bin/rild -l /system/lib/libxgold-ril.so -- -d /dev/xmd-tty0
     class main
     socket rild stream 660 root radio
     socket rild-debug stream 660 radio system
     user root
-    group radio cache inet misc audio sdcard_r sdcard_rw qcom_oncrpc qcom_diag log
+    group radio cache inet misc audio sdcard_r sdcard_rw bluetooth log system
 
 service surfaceflinger /system/bin/surfaceflinger
     class main
@@ -524,7 +525,7 @@ service drm /system/bin/drmserver
 service media /system/bin/mediaserver
     class main
     user media
-    group audio camera inet net_bt net_bt_admin net_bw_acct drmrpc qcom_diag
+    group audio camera inet net_bt net_bt_admin net_bw_acct drmrpc bluetooth
     ioprio rt 4
 
 service bootanim /system/bin/bootanimation
diff --git a/rootdir/ueventd.rc b/rootdir/ueventd.rc
index cee3e73..7c7f1d5 100644
--- a/rootdir/ueventd.rc
+++ b/rootdir/ueventd.rc
@@ -29,8 +29,8 @@
 /dev/android_adb          0660   adb        adb
 /dev/android_adb_enable   0660   adb        adb
 /dev/ttyMSM0              0600   bluetooth  bluetooth
-/dev/uhid                 0660   system     net_bt_stack
-/dev/uinput               0660   system     net_bt_stack
+/dev/uhid                 0660   system     bluetooth
+/dev/uinput               0660   system     bluetooth
 /dev/alarm                0664   system     radio
 /dev/tty0                 0660   root       system
 /dev/graphics/*           0660   root       graphics
