diff --git a/services/surfaceflinger/SurfaceFlinger.cpp b/services/surfaceflinger/SurfaceFlinger.cpp
index f31bae4..1442c69 100644
--- a/services/surfaceflinger/SurfaceFlinger.cpp
+++ b/services/surfaceflinger/SurfaceFlinger.cpp
@@ -2714,8 +2714,8 @@ status_t SurfaceFlinger::captureScreenImplLocked(const sp<IBinder>& display,
     }
 
     if ((sw > hw_w) || (sh > hw_h)) {
-        ALOGE("size mismatch (%d, %d) > (%d, %d)", sw, sh, hw_w, hw_h);
-        return BAD_VALUE;
+        sw = hw_w;
+        sh = hw_h;
     }
 
     sw = (!sw) ? hw_w : sw;
