--- pathmap.mk_	2013-05-14 12:21:45.000000000 +0800
+++ pathmap.mk	2013-05-14 12:21:45.000000000 +0800
@@ -27,6 +27,9 @@
 # A mapping from shorthand names to include directories.
 #
 pathmap_INCL := \
+    bluedroid:system/bluetooth/bluedroid/include \
+    bluez:external/bluetooth/bluez \
+    glib:external/bluetooth/glib \
     bootloader:bootable/bootloader/legacy/include \
     camera:system/media/camera/include \
     corecg:external/skia/include/core \
